src/wkt/src/float.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2025 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// A helper to serialize `f32` to ProtoJSON format.
+// TODO(#1767): use float_serialize from generated code.
+#[allow(dead_code)]
+fn float_serialize<S>(x: &f32, s: S) -> std::result::Result<S::Ok, S::Error>
+where
+    S: serde::ser::Serializer,
+{
+    // Handle special strings, see https://protobuf.dev/programming-guides/json/.
+    if x.is_nan() {
+        return s.serialize_str("NaN");
+    }
+    if x.is_infinite() {
+        if x.is_sign_negative() {
+            return s.serialize_str("-Infinity");
+        }
+        return s.serialize_str("Infinity");
+    }
+    s.serialize_f32(*x)
+}
+
+#[allow(dead_code)]
+struct FloatVisitor;
+
+/// A helper to deserialize `f32` from ProtoJSON format.
+// TODO(#1767): use float_serialize from generated code.
+#[allow(dead_code)]
+fn float_deserialize<'de, D>(deserializer: D) -> std::result::Result<f32, D::Error>
+where
+    D: serde::de::Deserializer<'de>,
+{
+    deserializer.deserialize_any(FloatVisitor)
+}
+
+impl serde::de::Visitor<'_> for FloatVisitor {
+    type Value = f32;
+
+    fn visit_str<E>(self, value: &str) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // Handle special strings, see https://protobuf.dev/programming-guides/json/.
+        match value {
+            "NaN" => Ok(f32::NAN),
+            "Infinity" => Ok(f32::INFINITY),
+            "-Infinity" => Ok(f32::NEG_INFINITY),
+            _ => Err(serde::de::Error::invalid_value(
+                serde::de::Unexpected::Other(value),
+                &"a valid ProtoJSON string for f32 (NaN, Infinity, -Infinity)",
+            )),
+        }
+    }
+
+    fn visit_f32<E>(self, value: f32) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        // TODO: Find a way to test this code path, serde_json floats
+        // stored as f64.
+        Ok(value)
+    }
+
+    // Floats in serde_json are f64.
+    fn visit_f64<E>(self, value: f64) -> std::result::Result<Self::Value, E>
+    where
+        E: serde::de::Error,
+    {
+        Ok(value as f32)
```

Comment:
This is correct, but I had to look it up:

https://doc.rust-lang.org/reference/expressions/operator-expr.html#r-expr.as.numeric.float-narrowing

Maybe a comment?

Reply:
Done.